torch/_dynamo/backends/distributed.py

```diff
@@ -332,32 +316,54 @@ def compile_fn(self, gm: fx.GraphModule, example_inputs: List[torch.Tensor]):
         debug_str += "\n---------------\n"
         ddp_graph_log.debug(debug_str)
 
-        # 3: compile each of the partitioned submodules using the user-provided compiler
-        class SubmodCompiler(torch.fx.interpreter.Interpreter):
+        # 3: Replace submodules with lazily compiling submodule
+        class SubmoduleReplacer(torch.fx.interpreter.Interpreter):
             def __init__(self, module, compiler):
                 super().__init__(module)
                 self.compiler = compiler
 
-            def compile_submod(self, input_mod, args, kwargs):
+            def lazily_compiled_submod(self, input_mod):
                 """
-                Compile the submodule,
-                using a wrapper to make sure its output is always a tuple,
-                which is required by AotAutograd based compilers
+                Create a wrapper around submodules which:
+                - lazily compiles each of the partitioned submodules using the user-provided compiler
+                - unpacks singleton tuples/lists into flat arg
                 """
-                assert len(kwargs) == 0, "We assume only args for these modules"
 
-                class WrapperModule(torch.nn.Module):
-                    def __init__(self, submod, unwrap_singleton_tuple):
+                class LazilyCompiledModule(torch.nn.Module):
+                    def __init__(self, submod, compiler, unwrap_singleton_tuple):
                         super().__init__()
                         self.submod = submod
+                        self.compiler = compiler
+                        self.compiled = False
                         self.unwrap_singleton_tuple = unwrap_singleton_tuple
 
                     def forward(self, *args):
+                        if not self.compiled:
+                            assert (
+                                fake_mode
+                            ), "fake mode must have been available when creating lazy submod"
+                            fake_args = []
+                            for arg in args:
+                                if isinstance(arg, torch.Tensor) and not is_fake(arg):
+                                    fake_args.append(
+                                        torch._dynamo.utils.to_fake_tensor(
+                                            arg, fake_mode
+                                        )
+                                    )
+                                else:
+                                    fake_args.append(arg)
+                            # First trace with fake args
+                            new_submod = self.compiler(self.submod, tuple(fake_args))
```

Comment:
```suggestion
new_submod = self.compiler(self.submod, args)
```
@jon-chuang If we make this change to use real tensors to do the compile, the internal test will also pass.

Reply:
Hmm I see. I just want to double check that we will not run the real computation on the real args. I think it is fine as we normally already pass real example inputs to our compilers.

Reply:
I guess there's no simple repro to add to the testing path? It would probably be better if we can find one, but I get that it's probably also hard to reproduce.

Reply:
@yf225 would you mind importing this PR (including new commits) into the meta CI? Just to confirm everything is solid before we merge again

Reply:
@jon-chuang I imported and tested it internally, and can confirm that everything works. We are ready to merge this PR.